File: src/main/java/org/flossware/civilization/model/ReligionState.java
package org.flossware.civilization.model;

import java.util.List;
import java.util.Map;

/**
 * Immutable religion and culture state.
 */
public record ReligionState(
    Map<String, Double> religionShares,
    double religiousUnity,
    double stabilityBonus,
    double spreadRate
) {
    public ReligionState {
        if (religiousUnity < 0 || religiousUnity > 1) {
            throw new IllegalArgumentException("Religious unity must be in [0, 1]");
        }
        if (stabilityBonus < 0 || stabilityBonus > 1) {
            throw new IllegalArgumentException("Stability bonus must be in [0, 1]");
        }
        religionShares = Map.copyOf(religionShares);
    }

    public ReligionState withReligionShares(Map<String, Double> newShares) {
        double totalShare = newShares.values().stream().mapToDouble(d -> d).sum();
        double newUnity = newShares.values().stream().mapToDouble(d -> d).max().orElse(0.0);
        double newBonus = newUnity * 0.2;
        return new ReligionState(newShares, newUnity, newBonus, spreadRate);
    }

    public ReligionState withSpreadRate(double newRate) {
        return new ReligionState(religionShares, religiousUnity, stabilityBonus, newRate);
    }
}

File: mobile/shared/src/commonMain/kotlin/org/flossware/civilization/model/ReligionState.kt
package org.flossware.civilization.model

import kotlinx.serialization.Serializable

@Serializable
data class ReligionState(
    val religionShares: Map<String, Double>,
    val religiousUnity: Double,
    val stabilityBonus: Double,
    val spreadRate: Double
) {
    init {
        require(religiousUnity in 0.0..1.0) { "Religious unity must be in [0, 1]" }
        require(stabilityBonus in 0.0..1.0) { "Stability bonus must be in [0, 1]" }
    }

    fun withReligionShares(newShares: Map<String, Double>): ReligionState {
        val newUnity = newShares.values.maxOrNull() ?: 0.0
        val newBonus = newUnity * 0.2
        return copy(religionShares = newShares, religiousUnity = newUnity, stabilityBonus = newBonus)
    }

    fun withSpreadRate(newRate: Double) = copy(spreadRate = newRate)
}
